### PR TITLE
De-event-sourcing — Phase A: brief, plan, and ADR 0001

### PR DESCRIPTION
**Pull request by alexander-yevsyukov**
## What this is

Phase A (design and planning) of migrating Spine `Aggregate`s off event
sourcing — loading aggregates from their latest persisted state instead of
replaying an event journal. Implementation follows in later PRs (Phase B onward).

**This PR also refreshes the repo's local Spine SDK dependencies** (see
*Dependency refresh* below) so future branches build on the latest snapshots —
the aggregate cutover will land on a current dependency baseline.

## Contents

- **[Brief](.agents/tasks/de-event-sourcing-brief.md)** — the problem: event
  sourcing is fragile (un-loadable aggregates, partial validity, un-retirable
  history) and the high-level solution.
- **[Delivery plan](.agents/tasks/de-event-sourcing-plan.md)** — phased
  execution plan (A design → B core cutover in 3 PRs → C storage/SPI → D state
  history → E downstream rollout → F org-wide `@Apply` gate).
- **[ADR 0001](docs/adr/0001-aggregates-without-event-sourcing.md)** —
  **Accepted.** Resolves the eight design decisions, each signed off after an
  individual walkthrough. Review follow-ups (2026-07-04/05) added **D9–D10**
  and **revised D1** (event import dropped, with the usage research recorded
  in the ADR).

## Accepted decisions (ADR 0001)

- **D1** *(revised 2026-07-05)* — **event import is dropped**: no `@Import`
  receptor; `ImportBus`, import routing, and `BlackBox.importsEvent` are
  removed in PR-B2; `InboxLabel.IMPORT_EVENT` and the `EventImported` system
  event stay deprecated for wire compatibility.
- **D2** — `@Apply` becomes a fail-fast `ModelError` with a self-contained
  migration message; annotation deprecated until v2.0.0 (final).
- **D3** — aggregate version advances **+1 per command** (process-manager
  semantics); event ordering rides the existing per-event microsecond timestamps.
- **D4** — state update is never forced in any handler; `@React` emission
  optional; the store gate must also fire on a business-state change.
- **D5** — deduplication moves to the delivery layer; the aggregate
  `IdempotencyGuard` becomes **opt-in, off by default**; recent history loads
  **lazily on demand**, bounded by `historyDepth` (default 100).
- **D6** — open the transaction **before** the handler; single validate-at-commit;
  rollback discards the whole dispatch (fixes partial validity).
- **D7** — defer journal trimming to Phase D; journal append-only in between.
- **D8** — relax the applier-only state-access guard.
- **D9** *(added in review)* — preventive validation: `tryAlter {}`
  (validate-before-apply on a scratch builder) plus the
  `ValidatingBuilder.validate()` in-place probe (lands in the `validation`
  repository).
- **D10** *(added in review)* — explicit recent-history window:
  `historyBackward(depth)` / `historyContains(depth, predicate)`; the
  parameterless forms are deprecated, defaulting to `historyDepth`.

## Dependency refresh

Local Spine SDK snapshot dependencies bumped to their latest published versions,
with the `docs/dependencies` reports regenerated. `./gradlew build` passes with
the updated set.

| Object | Old → New |
|--------|-----------|
| `Base` | `.421` → `.423` |
| `CoreJvm` | `.381` → `.383` |
| `Logging` | `.419` → `.422` |
| `Time` | `.242` → `.244` |
| `ToolBase` | `.402` → `.403` |

`ToolBase.dogfoodingVersion` (build-time plugin) left at `.402` intentionally;
the other 10 `local/` objects were already current.

## Notes

- Version bumped to `2.0.0-SNAPSHOT.400`.
- Next PR is **PR-B1** (additive preventive validation: `tryAlter`).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by alexander-yevsyukov** [comment]
@armiol, PTAL at ADR in this PR. 

If we do need to change anything in the plan, let's settle this before we go with the implementation.

**Comment by alexander-yevsyukov** [comment]
@codex review.

File: docs/adr/0001-aggregates-without-event-sourcing.md [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+| Receptor | Contract |
+|----------|----------|
+| `@Assign` (command) | **May** update state via `builder()` and/or set lifecycle flags via `setArchived()` / `setDeleted()` — state update is optional. **Must emit ≥1 event or reject** — empty success stays illegal. |
+| `@React` (reaction) | Same as `@Assign`, **except emitting an event is optional** — a reactor may emit zero events. May update state and set lifecycle flags, or neither. `AggregateEventReactionEndpoint.onEmptyResult` stays a no-op (`AggregateEventReactionEndpoint.java:60-67`). |
+| `@Import` | Consumes the imported event and **may** update state via `builder()`, or leave it unchanged — **both are OK, neither is an error.** The imported event is always recorded as the fact and the version advances +1 (D1). |
+
+`setArchived()` / `setDeleted()` are explicitly allowed in `@Assign` and `@React`
+bodies. The store decision is unchanged: `AggregateEndpoint.storeAndPost`
+persists when `withEvents || lifecycleFlags changed` (`AggregateEndpoint.java:83-100`).
+
```
[the remark this answers is not in the document]
Comment (alexander-yevsyukov) [reply docs/adr/0001-aggregates-without-event-sourcing.md:281]: Good catch — this was a real gap, not just wording. Fixed in 66fdbbcaee: D4 now states the store condition must be **expanded** to also persist when the built business state changed (`withEvents || lifecycleFlags changed || state changed`). Since a zero-event `@React` can now mutate state directly, the old gate would have silently dropped it. PR-B2 is required to implement this and add a test for a zero-event, state-only reactor; mirrored in the delivery plan's save-path step 9.

File: docs/adr/0001-aggregates-without-event-sourcing.md [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+---
+
+## Decision summary
+
+| # | Decision |
+|---|----------|
+| D1 | Event import survives via a new Kotlin `@Import` receptor that mutates `builder()`; the imported event is recorded as the fact. `ImportBus`/routing unchanged. |
+| D2 | `@Apply` on any aggregate (incl. `AggregatePart`) is a **`ModelError` at model-building time** after cutover — fail fast, never a silent no-op. |
+| D3 | Aggregate version advances **+1 per command dispatch**, not per event — `ProcessManager` semantics via `CommandDispatchingPhase` + `VersionIncrement.sequentially`. |
+| D4 | The "must produce an event" guard is **endpoint-scoped**: `@Assign` must emit ≥1 event; `@React` may be a no-op or a lifecycle-only change; `@Import` must mutate state and needs no emitted event. |
```
[the remark this answers is not in the document]
Comment (alexander-yevsyukov) [reply docs/adr/0001-aggregates-without-event-sourcing.md]: Fixed in 66fdbbcaee. The D1 and D4 summary rows now say `@Import` *may* mutate state (or leave it unchanged), matching the approved D1/D4 body — an import that only journals an external fact without changing state is valid and must not be rejected.

**Comment by alexander-yevsyukov** [comment]
Added a refresh of the local Spine SDK dependencies (Base, CoreJvm, Logging, Time, ToolBase → latest snapshots) plus regenerated `docs/dependencies` reports, in commit d2989ab08e. `./gradlew build` passes locally. This gives the aggregate cutover a current dependency baseline; the PR is no longer docs-only.

**Review by armiol — CHANGES_REQUESTED** [review]
@alexander-yevsyukov 
So far I have several questions. 

All of them should be ideally resolved by changing/appending the planning artifacts with the precise code samples of the future `core-jvm` Aggregates used in some test app + comments with explanation.

1. How do we give developers a control over the state becoming invalid — and potentially changing the behavior? In simple cases a corresponding exception is going to be thrown directly on builder's "setter" IIRC. But it will not be the case with complex or custom validation rules.
2. What exactly happends to `React`-ors in Aggregates? Basically, whatever we do in p.1, we should do here.
3. What happens to `@Import`-annotated receptors (yes, I read everything there is in this PR)? How do we change version of the Aggregate? 

There is also a small question regarding `recentHistory`. This call will need to have a parameter now, because we no longer load "everything up until the last snapshot".

**Comment by alexander-yevsyukov** [comment]
@armiol Thank you for the review! All four points are addressed on the branch:
the ADR gained two new decisions — **D9** and **D10** — each designed and
approved after an individual walkthrough, with code samples in the *Public API
sketch* section. Point by point:

**1. Control over the state becoming invalid.** Resolved by new
[D9 — Preventive state validation in receptors](https://github.com/SpineEventEngine/core-jvm/blob/de-event-sourcing/docs/adr/0001-aggregates-without-event-sourcing.md#d9--preventive-state-validation-in-receptors-tryalter).
Two parts:

- The simple case — a throwing setter (today only `(set_once)`) — now surfaces
  *inside the handler body*, because under D6 handlers mutate `builder()`
  directly. It can be caught right at the faulty line.
- For build-time constraints (`(required)`, `(min)`, custom rules, …) there is
  a new `tryAlter {}`: it runs the mutation on a **scratch** copy of the
  builder, validates the candidate via the same path the commit uses, and
  merges into the live builder only when clean — otherwise it returns the
  `ConstraintViolation`s and the live state stays untouched:

  ```kotlin
  @Assign
  @Throws(OutOfStock::class)
  fun handle(cmd: ReserveItems): ItemsReserved {
      val violations = tryAlter { inStock = state.inStock - cmd.quantity }
      if (violations.isNotEmpty()) {
          throw OutOfStock.newBuilder()./* … */.build()   // business NACK
      }
      return itemsReserved { /* … */ }
  }
  ```

  Setter-thrown `ValidationException`s inside the block are folded into the
  returned list, so both constraint families surface at one control point. A
  receptor that skips `tryAlter` keeps today's semantics: single
  validate-at-commit, rollback, failed dispatch — D6 remains the safety net.

**2. `@React`-ors.** Exactly "whatever we do in p. 1": `tryAlter` is defined
for `TransactionalEntity`, so reactors (and process managers, which have had
this same gap all along) get the identical tool. Only the failure vocabulary
differs: a command handler **rejects**, while a reactor **skips the update**
and typically returns `NoReaction` — which speaks only about emission (D4);
the state stays untouched because `tryAlter` withheld the change, not because
`NoReaction` was returned. Sample in the API sketch.

**3. `@Import` and the version.** Covered by D1 + D3: an `@Import` receptor
may mutate `builder()` or no-op, and returns nothing — the imported event *is*
the recorded fact, which the endpoint injects into the outcome as the produced
event. The aggregate version advances **+1 per dispatch** (`ProcessManager`
semantics, D3), and for import it advances whether or not the state changed
(D1, *Consequences*). An import sample — including a loud `tryAlter` failure —
is in the API sketch as well.

**4. `recentHistory` parameter.** Agreed — resolved by new
[D10 — Explicit recent-history window](https://github.com/SpineEventEngine/core-jvm/blob/de-event-sourcing/docs/adr/0001-aggregates-without-event-sourcing.md#d10--explicit-recent-history-window-historybackwarddepth):
`historyBackward(int depth)` / `historyContains(int depth, Predicate)` state
the window at the call site — up to `depth` latest journal events, newest
first. The parameterless forms are deprecated but keep working, delegating
with `depth = historyDepth` (default 100 — the old `DEFAULT_SNAPSHOT_TRIGGER`),
and `historyDepth` itself is demoted to the `IdempotencyGuard`'s window. The
migration guide will call out the window change ("since the last snapshot" →
"the last `historyDepth` events").

On the test-app ask: the precise samples with comments now live in the ADR's
*Public API sketch* (all three receptor kinds); runnable fixtures follow in
PR-B1 (the `tryAlter` test suite) and PR-B2 (the fixture migration).

Commits: ca83d77a1b (D9), ed4e998ae6 (D10).

**Comment by alexander-yevsyukov** [comment]
@armiol, there's also another (additional) option to the suggested `tryAlter` approach. It is to add `builder().validate()` method that would return `List<ConstraintViolation>`. It would allow to validate a builder which already may have been corrupted.

It can be done by adding `default` interface methods to `ValidatingBuilder`. We have `buildPartial()` and `ValidatableMessage.validate()` in our hands.

**Comment by armiol** [comment]
> @armiol, there's also another (additional) option to the suggested `tryAlter` approach. It is to add `builder().validate()` method that would return `List<ConstraintViolation>`. It would allow to validate a builder which already may have been corrupted.
> 
> It can be done by adding `default` interface methods to `ValidatingBuilder`. We have `buildPartial()` and `ValidatableMessage.validate()` in our hands.

Let's see that in this plan (for some reason I cannot mention Claude).

**Comment by armiol** [comment]
@alexander-yevsyukov I also think that event import may be a conceptually wrong feature now, because Aggregates are no longer "events instead of a state". 

Or at least, with this name — where do we `@Import` to? 

I also cannot remember why we have it and if anyone has used it in production. 

I suggest we (Claude?) do a small research within this planning task on whether there is a functional gap covered by `@Import`, or end-users may just use `@React`.

**Comment by alexander-yevsyukov** [comment]
@armiol Captured on the branch: D9 in ADR 0001 now carries a dated amendment —
`default ValidatingBuilder.validate(): List<ConstraintViolation>` (built from
`buildPartial()` + `ValidatableMessage.validate()`), with the division of
labor spelled out: `validate()` probes a builder in place — including one that
already holds invalid content — while `tryAlter` protects the live builder
(reactors/imports). The delivery plan gained the cross-repo item (PR-B1,
step 7): the method lands in the `validation` repo and `core-jvm` consumes it
via a dependency bump; `tryAlter` does not block on it. Commit: 9de92c8fd0.

**Comment by alexander-yevsyukov** [comment]
@armiol Research done — and the decision is made: **event import is dropped**
(ADR D1, revised 2026-07-05). The findings, now recorded inside D1:

- Import exists since 2018 for two scenarios (`ImportBus` Javadoc): adopting
  facts of a legacy/third-party system as the context's own history, and
  recording facts without intermediate commands or events. Defining mechanics:
  unicast delivery to one aggregate + adoption (journaled as the aggregate's
  *own* fact, re-stamped with its version, then posted).
- Usage: production — **one class, ever**: `GoogleGroupPart` in the dormant
  `auth` repository (Google Directory facts; last push 2020-03). Nothing in
  `spine-examples`, no downstream `setupImportRouting` or
  `BlackBox.importsEvent` callers; the rest is core-jvm's own fixtures.
- Functional gap vs `@React`: the replay-era substance — imported events
  *rebuilding state* — dies with the cutover. What remains (journal
  provenance, skipping intermediate messages) has standard replacements:
  `(external) = true` reactions, gateways/commands, storage-level seeding,
  and real signals in tests.

So: no `@Import` receptor. PR-B2 removes `ImportBus`, the import
endpoint/routing, and `BlackBox.importsEvent`; `InboxLabel.IMPORT_EVENT` and
the `EventImported` system event stay deprecated for wire compatibility. Your
naming question dissolves together with the feature 🙂 Commit: 19ddf8db99.

**Review by armiol — APPROVED** [review]
